### PR TITLE
[MXNET-433] Tutorial on saving and loading gluon models

### PR DESCRIPTION
**Pull request by indhub**
## Description ##
Tutorial on saving and loading gluon models

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

### CODE REVIEW

File: docs/tutorials/gluon/save_load_params.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+# MNIST images are 28x28. Total pixels in input layer is 28x28 = 784
+num_inputs = 784
+# Clasify the images into one of the 10 digits
+num_outputs = 10
+# 64 images in a batch
+batch_size = 64
+
+# Helper to preprocess data for training
+def transform(data, label):
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Would be better to use the new transforms. Something like `train_data = gluon.data.vision.MNIST(train=True).transform_first(transforms.ToTensor())`

File: docs/tutorials/gluon/save_load_params.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    return nd.transpose(data.astype(np.float32), (2,0,1))/255, label.astype(np.float32)
+
+# Load the training data
+train_data = gluon.data.DataLoader(gluon.data.vision.MNIST(train=True, transform=transform), 
+                                   batch_size, shuffle=True)
+
+# Build a simple convolutional network
+def build_lenet(net):
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Why does this need to take an argument? Can't we just create net inside of function here?

File: docs/tutorials/gluon/save_load_params.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+```python
+import mxnet as mx
+import mxnet.ndarray as nd
+from mxnet import nd, autograd, gluon
+
+import numpy as np
+```
+
+## Build and train a simple model
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Setup: build and train a simple model

File: docs/tutorials/gluon/save_load_params.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+## Loading model parameters from file
+
+Let's now create a network with the parameters we saved into the file. We build the network again using the helper first and then load the weights from the file we saved using the `load_params` function.
+
+```python
+new_net = build_lenet(gluon.nn.Sequential())
+new_net.load_params(file_name, ctx=ctx)
+```
+
+Note that to do this, we need the definition of the network as Python code. If our network is [Hybrid](https://mxnet.incubator.apache.org/tutorials/gluon/hybrid.html), we can even save the network architecture into files and we won't need the network definition in a Python file to load the network. We'll see how to do it in the next section.
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Would make a little more explicit that if on a different machine, you'd need to import the same function and run to create the same object before loading params.

File: docs/tutorials/gluon/save_load_params.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+        break
+
+verify_loaded_model(new_net)
+```
+![Model inputs](https://raw.githubusercontent.com/indhub/web-data/4a9c100aa996df3dff0e7f493029d411c2b526c3/mxnet/tutorials/gluon/save_load_params/mnist_in_1.png)
+
+Model predictions:  [1. 1. 4. 5. 0. 5. 7. 0. 3. 6.] <!--notebook-skip-line-->
+
+## Saving model architecture and weights to file
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Saving model parameters AND architecture to file

So that it matches with title format above.

File: docs/tutorials/gluon/save_load_params.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+We now have a trained hybrid network. This can be exported into files using the `export` function. The `export` function will export the model architecture into a `.json` file and model parameters into a `.params` file.
+
+```python
+net.export("lenet", epoch=1)
+```
+
+That's it! `export` in this case creates `lenet-symbol.json` and `lenet-0001.params` in the current directory.
+
+## Loading saved model architecture and weights from a different frontend
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Loading model parameters AND architecture from file

So that it matches with title format above. As major heading. Then create subheadings for 'from different frontend' and 'from Python'

File: docs/tutorials/gluon/save_load_params.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
+# Saving and Loading Gluon Models
+
+Training large models take a lot of time and it is a good idea to save the trained models to files to avoid training them again and again. There is a number of reasons to do this. For example, you might want to do inference on a machine that is different from the one where the model was trained. Sometimes model's performance on validation set decreases towards the end of the training because of overfitting. If you saved your model parameters after every epoch, at the end you can decide to use the model that performs best on the validation set.
+
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Would introduce that reader will be looking at two methods: params only and params and architecture. And would try to mention somewhere which method is recommended for certain situations. Currently not really discussed.

File: docs/tutorials/gluon/save_load_params.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+## Loading saved model architecture and weights from Python
+
+Serialized Hybrid networks (saved as .JSON and .params file) can be loaded and used inside Python frontend using `mx.model.load_checkpoint` and `gluon.nn.SymbolBlock`. To demonstrate that, let's load the network we serialized above.
+
+```python
+# Load the network architecture and parameters
+sym, arg_params, aux_params = mx.model.load_checkpoint('lenet', 1)
+# Create a Gluon Block using the loaded network architecture
+deserialized_net = gluon.nn.SymbolBlock(outputs=sym, inputs=mx.sym.var('data'))
```
Comment (thomelane) [inline docs/tutorials/gluon/save_load_params.md]: Would be good to explain data here. Where does the name come from?

File: docs/tutorials/index.md [hunk 9]
```diff
@@ -38,7 +38,7 @@ Select API:&nbsp;
     * [Visual Question Answering](http://gluon.mxnet.io/chapter08_computer-vision/visual-question-answer.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
 * Practitioner Guides
     * [Multi-GPU training](http://gluon.mxnet.io/chapter07_distributed-learning/multiple-gpus-gluon.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
-    * [Checkpointing and Model Serialization (a.k.a. saving and loading)](http://gluon.mxnet.io/chapter03_deep-neural-networks/serialization.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
+    * [Saving and Loading Models](/tutorials/gluon/save_load_params.html)
```
Comment (thomelane) [inline docs/tutorials/index.md]: Could leave straight dope article, and add as alternative link? Unless it's misleading, in which case we should submit pr for straight dope.

File: docs/tutorials/gluon/save_load_params.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+# Set the parameters
+net_params = deserialized_net.collect_params()
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/save_load_params.md]: recently learned you can load the parameters like that:
```
deserialized_net.collect_params().load('lenet-0001.params')
```
rather than:
````
net_params = deserialized_net.collect_params()
for param in arg_params:
    if param in net_params:
        net_params[param]._load_init(arg_params[param], ctx=ctx)
for param in aux_params:
    if param in net_params:
        net_params[param]._load_init(aux_params[param], ctx=ctx)
```

File: docs/tutorials/gluon/save_load_params.md (hunk 2 above)
Comment (indhub) [reply docs/tutorials/gluon/save_load_params.md]: So that, I can use this function to build either Block or HybridBlock. I'm building the network as a Block to demonstrate saving and loading parameters. I'm then building the network as HybridBlock to demonstrate saving and loading parameters and model architecture.

File: docs/tutorials/gluon/save_load_params.md (hunk 7 above)
Comment (indhub) [reply docs/tutorials/gluon/save_load_params.md]: Done.

File: docs/tutorials/gluon/save_load_params.md [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+1. [Loading serialized Hybrid networks from C](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/predict-cpp/image-classification-predict.cc)
+2. [Loading serialized Hybrid networks from Scala](https://github.com/apache/incubator-mxnet/blob/master/scala-package/infer/src/main/scala/org/apache/mxnet/infer/ImageClassifier.scala)
+
+### From Python
+
+Serialized Hybrid networks (saved as .JSON and .params file) can be loaded and used inside Python frontend using `mx.model.load_checkpoint` and `gluon.nn.SymbolBlock`. To demonstrate that, let's load the network we serialized above.
+
+```python
+# Load the network architecture and parameters
+sym, arg_params, aux_params = mx.model.load_checkpoint('lenet', 1)
```
Comment (ThomasDelteil) [inline docs/tutorials/gluon/save_load_params.md]: Since we just need `sym` I would suggest to rather use:
sym = mx.sym.load_json(open('lenet-symbol.json', 'r').read()

**Comment by thomelane** [comment]
Checked your changes @indhub . Good to be merged @szha.

File: docs/tutorials/gluon/save_load_params.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
+# Saving and Loading Gluon Models
+
+Training large models take a lot of time and it is a good idea to save the trained models to files to avoid training them again and again. There is a number of reasons to do this. For example, you might want to do inference on a machine that is different from the one where the model was trained. Sometimes model's performance on validation set decreases towards the end of the training because of overfitting. If you saved your model parameters after every epoch, at the end you can decide to use the model that performs best on the validation set.
```
Comment (sandeep-krishnamurthy) [inline docs/tutorials/gluon/save_load_params.md]: nit: There -are- a number of reasons..

Comment (sandeep-krishnamurthy) [reply docs/tutorials/gluon/save_load_params.md]: nit: another motivation would be to separate research from production by using more research native Python for training and Scala/C++ in production inference.

File: docs/tutorials/gluon/save_load_params.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+        net.add(gluon.nn.Dense(512, activation="relu"))
+        # Second fully connected layer with as many neurons as the number of classes
+        net.add(gluon.nn.Dense(num_outputs))
+        
+        return net
+
+# Train a given model using MNIST data
+def train_model(model):
+    # Initialize the parameters with Xavier initializer
+    net.collect_params().initialize(mx.init.Xavier(), ctx=ctx)
```
Comment (sandeep-krishnamurthy) [inline docs/tutorials/gluon/save_load_params.md]: model.collect_params?

net still works as it is a global parameter in your script.

File: docs/tutorials/gluon/save_load_params.md [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+    # Use cross entropy loss
+    softmax_cross_entropy = gluon.loss.SoftmaxCrossEntropyLoss()
+    # Use Adam optimizer
+    trainer = gluon.Trainer(net.collect_params(), 'adam', {'learning_rate': .001})
```
Comment (sandeep-krishnamurthy) [inline docs/tutorials/gluon/save_load_params.md]: same as above. net->model

File: docs/tutorials/gluon/save_load_params.md (hunk 13 above)
Comment (indhub) [reply docs/tutorials/gluon/save_load_params.md]: Good catch. thanks!
